docs(batch): explain record type discrepancy in failure and success handler (#2868)
awslabs/aws-lambda-powertools-python: docs/utilities/batch.md

For these scenarios, you can subclass `BatchProcessor` and quickly override `success_handler` and `failure_handler` methods:

> * **`success_handler()`** – is called for each successfully processed record
> * **`failure_handler()`** – is called for each failed record

Reviewer: ```suggestion
* **`success_handler()`** is called for each successfully processed record
* **`failure_handler()`** is called for each failed record
```

> ???+ warning

Reviewer: ```suggestion
???+ note
```

Less triggering for some readers/persona.

    These functions have a common `record` argument. For backward compatibility reasons, their type is not the same:

    - `success_handler`: `record` type is `dict[str, Any]`, the raw record data

>     - `failure_handler`: `record` type is a Pydantic model, either the parsed record or the  `EventSourceDataClass` record in case of validation error. Refer to [Accessing processed messages](#accessing-processed-messages) for more details

Reviewer: ```suggestion
    - `failure_handler`: `record` type can be an Event Source Data Class or your [Pydantic model](#pydantic-integration). During Pydantic validation errors, we fall back and serialize `record` to Event Source Data Class to not break the processing pipeline. Refer to [Accessing processed messages](#accessing-processed-messages) for more details.
```

Reviewer: Small correction since the default brings a Event Source Data Class unless you bring a Pydantic model. When the latter happens, we fall back to Event Source Data Class upon validation errors.

???+ example

> 	Let's suppose you'd like to add metrics tracking processing successes and failures of your batch records.

Reviewer: ```suggestion
	Let's suppose you'd like to add metrics to track successes and failures of your batch records.
```

Reviewer: typo

> ???+ example
> 	Let's suppose you'd like to add metrics to track successes and failures of your batch records.

Reviewer: ```suggestion
Let's suppose you'd like to add metrics to track successes and failures of your batch records.
```

Reviewer: Just noticed before I merged that we have two callouts, this can be noisy for some readers. The callout you just added is more important - I'll suggest moving to Note too so it's less of a danger-feel

<img width="1200" alt="image" src="https://github.com/aws-powertools/powertools-lambda-python/assets/3340292/70c643b3-0352-47c4-afec-7a36d287d524">

> ```python hl_lines="8 9 18-25 28 44" title="Extending failure handling mechanism in BatchProcessor"

Reviewer: ```suggestion
```python hl_lines="8-10 18-25 28 44" title="Extending failure handling mechanism in BatchProcessor"
```

This will also highlight `FailureResponse` type
